### PR TITLE
Jaelynlitz/refactor ndrop

### PR DESCRIPTION
**Pull request by jaelynlitz**
After discussion with @mjs271 and @odiazib, explmix should be refactored to only compute on a single level k (with k-1, and k+1 as well) then explmix will be called in a loop over nlev at a higher level. 

@odiazib has a WIP refactor of get_aer_num in the branch collab/ndrop

### CODE REVIEW

**Comment by jaelynlitz** [comment]
closing #146 in favor of this PR

File: src/mam4xx/ndrop.hpp [hunk 1]
```diff
@@ -22,87 +22,110 @@ class NDrop {
...
+  q = max(q, 0);
+} // end explmix
+
+// calculates maximum supersaturation for multiple
+// competing aerosol modes.
+// Abdul-Razzak and Ghan, A parameterization of aerosol activation.
+// 2. Multiple aerosol types. J. Geophys. Res., 105, 6837-6844.
+KOKKOS_INLINE_FUNCTION
+void maxsat(const Real zeta,   // [dimensionless]
+            const Real eta[4], // [dimensionless]
```
Comment (odiazib) [inline src/mam4xx/ndrop.hpp:83]: change` 4 `-> `AeroConfig::num_modes()`

File: src/mam4xx/ndrop.hpp [hunk 2]
```diff
@@ -22,87 +22,110 @@ class NDrop {
...
+        (dt *
+         (-src +
+          (ek_kp1 * (qold_kp1 - qold_k + (qactold_kp1 * (1 - overlap_kp1)))) +
+          (ek_km1 * (qold_km1 - qold_k + (qactold_km1 * (1 - overlap_km1))))));
+  } else {
+    q = qold_k + (dt * (src + (ek_kp1 * ((overlap_kp1 * qold_kp1) - qold_k)) +
+                        (ek_km1 * ((overlap_km1 * qold_k) - qold_k))));
+  }
+  // force to non-negative
+  q = max(q, 0);
```
Comment (odiazib) [inline src/mam4xx/ndrop.hpp:74]: I would use `haero:max` here.

File: src/mam4xx/ndrop.hpp [hunk 3]
```diff
@@ -22,87 +22,110 @@ class NDrop {
...
+
+// calculates maximum supersaturation for multiple
+// competing aerosol modes.
+// Abdul-Razzak and Ghan, A parameterization of aerosol activation.
+// 2. Multiple aerosol types. J. Geophys. Res., 105, 6837-6844.
+KOKKOS_INLINE_FUNCTION
+void maxsat(const Real zeta,   // [dimensionless]
+            const Real eta[4], // [dimensionless]
+            const Real nmode,  // modes
+            const Real smc[4], // critical supersaturation for number mode
```
Comment (odiazib) [inline src/mam4xx/ndrop.hpp:85]: change` 4 `-> `AeroConfig::num_modes()`

File: src/mam4xx/ndrop.hpp [hunk 4]
```diff
@@ -22,87 +22,110 @@ class NDrop {
...
+
+  Real sum = 0;
+  Real g1, g2;
+  bool weak_forcing = true; // whether forcing is sufficiently weak or not
+
+  for (int m = 0; m < nmode; m++) {
+    if (zeta > 1e5 /*FIXME: BAD CONSTANT*/ * eta[m] ||
+        smc[m] * smc[m] > 1e5 /*FIXME: BAD CONSTANT*/ * eta[m]) {
+      // weak forcing. essentially none activated
+      smax = 1e-20; /*FIXME: BAD CONSTANT*/
```
Comment (odiazib) [inline src/mam4xx/ndrop.hpp:101]: We could define a variable for this  numerical constant, i.e., `Real const small = 1e-20;`

File: src/mam4xx/ndrop.hpp [hunk 5]
```diff
@@ -22,87 +22,110 @@ class NDrop {
...
+  }
+
+  if (weak_forcing)
+    return;
+
+  for (int m = 0; m < nmode; m++) {
+    f1[m] = 0.5 * haero::exp(2.5 * haero::log(modes(m).mean_std_dev) *
+                             haero::log(modes(m).mean_std_dev));
+    f2[m] = 1.0 + 0.25 * haero::log(modes(m).mean_std_dev);
+    if (eta[m] > 1e-20 /*FIXME: BAD CONSTANT*/) {
```
Comment (odiazib) [inline src/mam4xx/ndrop.hpp:116]: we could this:` eta[m] > small`.

File: src/mam4xx/ndrop.hpp [hunk 6]
```diff
@@ -22,87 +22,110 @@ class NDrop {
...
+      g1 = (zeta / eta[m]) * haero::sqrt(zeta / eta[m]);
+      g2 = (smc[m] / haero::sqrt(eta[m] + 3.0 * zeta)) *
+           haero::sqrt(smc[m] / haero::sqrt(eta[m] + 3.0 * zeta));
+      sum += (f1[m] * g1 + f2[m] * g2) / (smc[m] * smc[m]);
+    } else {
+      sum = 1e20; /*FIXME: BAD CONSTANT*/
```
Comment (odiazib) [inline src/mam4xx/ndrop.hpp:122]: `sum=1e20 `-> `sum = 1./small`;

**Review by odiazib — APPROVED** [review]
I aded a few comments. It looks great!

File: src/mam4xx/ndrop.hpp (hunk 4 above)
Comment (jeff-cohere) [reply src/mam4xx/ndrop.hpp:101]: This constant could go at the top of the function.

File: src/mam4xx/ndrop.hpp (hunk 6 above)
Comment (jeff-cohere) [reply src/mam4xx/ndrop.hpp:122]: Or we could even have another constant `const Real big = 1/small;`

File: src/tests/mam4_ndrop_unit_tests.cpp [hunk 7]
```diff
@@ -140,70 +137,105 @@ TEST_CASE("test_get_aer_num", "mam4_ndrop") {
...
+  ekat::Comm comm;
+  ekat::logger::Logger<> logger("ndrop maxsat unit tests",
+                                ekat::logger::LogLevel::debug, comm);
+
+  logger.info("start of maxsat test");
+  int nmodes = AeroConfig::num_modes();
+  logger.info("nmodes = {}", nmodes);
+
+  Real zeta = 0;
+  Real eta[4];
```
Comment (jeff-cohere) [inline src/tests/mam4_ndrop_unit_tests.cpp:203]: These `4`s could be set to `nmodes`, which itself could be a `constexpr`.

**Review by jeff-cohere — APPROVED** [review]
This looks great. I have only a few minor comments and endorsements of @odiazib 's feedback. Thanks for reorganizing this in terms of single vertical levels.

**Review by pressel — CHANGES_REQUESTED** [review]
Thanks, @jaelynlitz this code looks good. Let's try to get some validation tests on this code before we merge. Let me know if I can help with that.

**Review by mjs271 — APPROVED** [review]
Looks great--nice work!

**Comment by jaelynlitz** [comment]
> Thanks, @jaelynlitz this code looks good. Let's try to get some validation tests on this code before we merge. Let me know if I can help with that.

Could we merge this branch to main then develop the validation testing in `collab/ndrop`? @odiazib already has a driver outlined there.

Alternatively, I could just merge this branch to `collab/ndrop` and continue dev there?

**Comment by jeff-cohere** [comment]
Hey @jaelynlitz ! Whenever you feel it's ready, please merge. Merging to `main` is fine with me if it's okay with you.

**Pull request merged**